cpp/src/arrow/array-binary-test.cc

```diff
@@ -522,6 +522,19 @@ TEST_F(TestChunkedBinaryBuilder, BasicOperation) {
   }
 }
 
+TEST_F(TestChunkedBinaryBuilder, Reserve) {
+  const int32_t chunksize = 1000;
+  Init(chunksize);
+  auto bytes_before_reserve = default_memory_pool()->bytes_allocated();
+  ASSERT_OK(builder_->Reserve(chunksize / 2));
+  for (int i = 0; i < 8; ++i) {
+    ASSERT_OK(builder_->Reserve(chunksize / 2));
+  }
+  // no new memory will be allocated since capacity was sufficient for the loop's
```

Comment:
How is that? We're reserving `4.5 * chunksize` total, no?

Reply:
In between Reserve calls, no data is appended to the builder, so there is no need to allocate any additional memory. 

If you look at the bug report in ARROW-6060 the issue appears to be that memory was being allocated when there was no need.

Reply:
To review the docstring for Reserve

```
  /// \brief Ensure that there is enough space allocated to add the indicated
  /// number of elements without any further calls to Resize. Overallocation is
  /// used in order to minimize the impact of incremental Reserve() calls.
  ///
  /// \param[in] additional_capacity the number of additional array values
  /// \return Status
```

So if there is already enough memory allocated then Reserve should be a no-op

Reply:
(it might be worth stating in the docstring that calls to Reserve are not cumulative)

Reply:
> (it might be worth stating in the docstring that calls to Reserve are not cumulative)

Really? They are. It's the (confusing) difference between Resize and Reserve.

Reply:
i.e. Reserve(100) should reserve allocated space for 100 _additional_ items.

Reply:
Yes, but in addition to what? For the current Reserve implementations the current length of the builder is considered (which for this test case remains 0, since I never Append anything)

Reply:
My intuition would be that Reserve ensure there's enough space for 100 items, if you call it without adding data, it should be a noop.

Reply:
Hmm, I see. Well I'd say that `Reserve` is really confusing then :-)

Reply:
@wesm I've added a note to the docstring

Reply:
This is a good opportunity to clarify the docstring if there is still confusion. 

`Reserve`: pre-allocate space for an _absolute, additional_ amount of data
`Resize`: pre-allocate space for an _absolute_ amount of data (regardless of how much has already been appended)

As we've seen here getting the behavior of Reserve wrong has made the 0.14.x release basically unusable for a number of users, so there are consequences to this

Reply:
@bkietz is it possible to have a more aggressive test, e.g. one that uses 16mb in the normal case and 16g in the falling one?

Reply:
@fsaintjacques in the failing case to balloon to 16G all I'd need to do is increase the [loop max](https://github.com/apache/arrow/pull/5016/files/9c5f9aad99067591ede26bc2ee06eb59dec14cca#diff-30e1e8083901d7b9b2abf6c430ff5c1fR530); the calls to Reserve would continue increasing capacity by the growth factor (1.5) until we run out of memory. I'm not sure what merit increasing the test's scale has

Reply:
Agreed with @bkietz. We don't need to test to crash, just fail, if `Reserve()` is buggy.